sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -327,12 +428,47 @@ DataSource buildDatasource() throws Exception{
         if (getConnectionProperties() != null && getConnectionProperties().get() != null) {
           basicDataSource.setConnectionProperties(getConnectionProperties().get());
         }
-        return basicDataSource;
+        current = basicDataSource;
       }
+
+      // wrapping the datasource as a pooling datasource
+      DataSourceConnectionFactory connectionFactory = new DataSourceConnectionFactory(current);
+      PoolableConnectionFactory poolableConnectionFactory =
+              new PoolableConnectionFactory(connectionFactory, null);
+      GenericObjectPoolConfig poolConfig = new GenericObjectPoolConfig();
+      poolConfig.setMaxTotal(getPoolMaxTotal());
+      poolConfig.setBlockWhenExhausted(getPoolBlockWhenExhausted());
+      poolConfig.setMaxWaitMillis(getPoolMaxWaitMillis());
+      poolConfig.setMaxIdle(getPoolMaxIdle());
+      poolConfig.setMinIdle(getPoolMinIdle());
+      poolConfig.setTestOnBorrow(getPoolTestOnBorrow());
+      poolConfig.setTestOnReturn(getPoolTestOnReturn());
+      poolConfig.setNumTestsPerEvictionRun(getPoolNumTestsPerEvictionRun());
+      poolConfig.setMinEvictableIdleTimeMillis(getPoolMinEvictableIdleTimeMillis());
+      poolConfig.setTestWhileIdle(getPoolTestWhileIdle());
+      poolConfig.setSoftMinEvictableIdleTimeMillis(getPoolSoftMinEvictableIdleTimeMillis());
+      poolConfig.setLifo(getPoolLifo());
+      GenericObjectPool connectionPool =
+              new GenericObjectPool(poolableConnectionFactory, poolConfig);
+      poolableConnectionFactory.setPool(connectionPool);
+      poolableConnectionFactory.setValidationQuery("SELECT 1 FROM DUAL");
```

Comment:
I don't think this works in all possible JDBC-compatible databases. I'm not sure we need a validation query in general. An invalid connection will lead to an error that will be simply retried anyway.

Reply:
My bad, I forgot to update with the provided value. By default, the `validationQuery` should be null and the user can define it depending of his database.